crates/glue/writing_a_glue_plugin.md

```diff
@@ -0,0 +1,441 @@
+# Writing a glue plugin
+
+In order to write a platform, the language the platform is written in must know how to interoperate with Roc. As Roc is statically typed, the required FFI wrapper code must be tailored to the datatypes that form the FFI boundry between the host and the Roc application. Handily, the `roc glue` subcommand will take these types as defined in a `platform.roc` file and turn them into layout information. A glue plugin then in turn uses this to generate the appropriate source code in a target language. This layout information comes in the form of a data structure that requires some additional explanation.
+
+- [Writing a glue plugin](#writing-a-glue-plugin)
+- [The Types data structure](#the-types-data-structure)
+- [The Shape data structure](#the-shape-data-structure)
+  - [Unit](#unit)
+  - [Bool](#bool)
+  - [Num](#num)
+  - [RocStr, RocList, RocDict, RocSet, RocBox](#rocstr-roclist-rocdict-rocset-rocbox)
+  - [Struct](#struct)
+    - [HasNoClosure](#hasnoclosure)
+    - [HasClosure](#hasclosure)
+  - [TagUnion](#tagunion)
+    - [NonRecursive](#nonrecursive)
+    - [Enumeration](#enumeration)
+    - [SingleTagStruct](#singletagstruct)
+    - [Recursive](#recursive)
+      - [With pointer tagging](#with-pointer-tagging)
+      - [Without pointer tagging](#without-pointer-tagging)
+    - [NullableWrapped](#nullablewrapped)
+    - [NullableUnwrapped](#nullableunwrapped)
+    - [NonNullableUnwrapped](#nonnullableunwrapped)
+  - [EmptyTagUnion](#emptytagunion)
+  - [TagUnionPayload](#tagunionpayload)
+  - [Function](#function)
+  - [Unsized](#unsized)
+  - [RecursivePointer](#recursivepointer)
+
+# The Types data structure
+
+`Types` is a data structure that stores the layout information of all types on the host-application FFI boundry for a specific architecture. This includes these types' size, alignment, order of declaration, and their `Shape`.
+
+# The Shape data structure
+
+`Shape` is where `roc glue` stores all layout information of a type specific to its kind. It consists of a tag union with the following tags:
+
+## Unit
+
+An empty record `{}`, Roc's anonymous zero-sized type.
+
+## Bool
+
+A one-byte type representing `true` as 1 and `false` as 0.
+
+## Num
+
+A number type. Either a signed integer (I8 through I128), an unsigned integer (U8 through U128), an IEEE 754 floating point number (F32 or F64), or a Roc-specific fixed-point number type ([Dec](https://www.roc-lang.org/builtins/Num#Dec)).
+
+## RocStr, RocList, RocDict, RocSet, RocBox
+
+Roc builtin datatypes. The Roc compiler is planned to expose parts of the standard library to glue code via C FFI to avoid having the reimplement them by hand in every host language. In the meantime, Rust implementations of the `Str`, `List`, and `Box` datatypes can be found under `crates/roc_std` for reference.
```

Comment:
Maybe mention that the source of truth is the zig implementation and link to it. Cause sometimes even the rust falls behind or has bugs.